File: apps/agents/app/services/realtime.py
from __future__ import annotations

from typing import Any
from uuid import uuid4

from app.clients.pusher_client import get_pusher_client


def workspace_channel(workspace_id: str) -> str:
    return f"private-workspace-{workspace_id}"


def conversation_channel(conversation_id: str) -> str:
    return f"private-conversation-{conversation_id}"


def _trigger(channel: str, event: str, payload: dict[str, Any]) -> None:
    try:
        client = get_pusher_client()
        client.trigger(channel, event, payload)
    except Exception:
        return


def emit_message_created(workspace_id: str, conversation_id: str, message: dict[str, Any]) -> None:
    payload = {
        "event_id": str(uuid4()),
        "workspace_id": workspace_id,
        "conversation_id": conversation_id,
        "message": message,
    }
    _trigger(conversation_channel(conversation_id), "message:created", payload)


def emit_attachment_created(
    workspace_id: str,
    conversation_id: str,
    message_id: str,
    attachment: dict[str, Any],
) -> None:
    payload = {
        "event_id": str(uuid4()),
        "workspace_id": workspace_id,
        "conversation_id": conversation_id,
        "message_id": message_id,
        "attachment": attachment,
    }
    _trigger(conversation_channel(conversation_id), "attachment:created", payload)


def emit_conversation_updated(
    workspace_id: str,
    conversation_id: str,
    updates: dict[str, Any],
) -> None:
    payload = {
        "event_id": str(uuid4()),
        "workspace_id": workspace_id,
        "conversation_id": conversation_id,
        **updates,
    }
    _trigger(workspace_channel(workspace_id), "conversation:updated", payload)


def emit_tags_updated(
    workspace_id: str,
    conversation_id: str,
    tags: list[str],
) -> None:
    payload = {
        "event_id": str(uuid4()),
        "workspace_id": workspace_id,
        "conversation_id": conversation_id,
        "tags": tags,
    }
    _trigger(workspace_channel(workspace_id), "tags:updated", payload)

File: apps/baileys/src/realtime.js
import crypto from "crypto";
import Pusher from "pusher";

export const createRealtime = ({
  logger,
  appId,
  key,
  secret,
  cluster,
}) => {
  let pusherServer = null;
  let pusherWarned = false;

  const workspaceChannel = (workspaceId) => `private-workspace-${workspaceId}`;
  const conversationChannel = (conversationId) =>
    `private-conversation-${conversationId}`;

  const getPusherServer = () => {
    if (!appId || !key || !secret || !cluster) {
      if (!pusherWarned) {
        pusherWarned = true;
        logger.warn(
          "Pusher envs ausentes; realtime do inbox sera desabilitado."
        );
      }
      return null;
    }
    if (!pusherServer) {
      pusherServer = new Pusher({
        appId,
        key,
        secret,
        cluster,
        useTLS: true,
      });
    }
    return pusherServer;
  };

  const triggerPusher = async (channel, event, payload) => {
    const pusher = getPusherServer();
    if (!pusher) return;
    try {
      await pusher.trigger(channel, event, payload);
    } catch (error) {
      logger.warn({ err: error, event, channel }, "Falha ao disparar Pusher");
    }
  };

  const emitMessageCreated = async ({ workspaceId, conversationId, message }) => {
    await triggerPusher(conversationChannel(conversationId), "message:created", {
      event_id: crypto.randomUUID(),
      emitted_at: new Date().toISOString(),
      workspace_id: workspaceId,
      conversation_id: conversationId,
      message,
    });
  };

  const emitConversationUpdated = async ({
    workspaceId,
    conversationId,
    status,
    lastMessage,
    lastAt,
    tags,
  }) => {
    await triggerPusher(workspaceChannel(workspaceId), "conversation:updated", {
      event_id: crypto.randomUUID(),
      emitted_at: new Date().toISOString(),
      workspace_id: workspaceId,
      conversation_id: conversationId,
      ...(status ? { status } : {}),
      ...(typeof lastMessage === "string"
        ? { ultima_mensagem: lastMessage }
        : {}),
      ...(lastAt ? { ultima_mensagem_em: lastAt } : {}),
      ...(Array.isArray(tags) ? { tags } : {}),
    });
  };

  const emitAttachmentCreated = async ({
    workspaceId,
    conversationId,
    messageId,
    attachment,
  }) => {
    await triggerPusher(conversationChannel(conversationId), "attachment:created", {
      event_id: crypto.randomUUID(),
      emitted_at: new Date().toISOString(),
      workspace_id: workspaceId,
      conversation_id: conversationId,
      message_id: messageId,
      attachment,
    });
  };

  return {
    emitMessageCreated,
    emitConversationUpdated,
    emitAttachmentCreated,
  };
};
